### ISSUE TITLE
Support Templates deferring package selection 

### ISSUE THREAD

**Issue by zentron** [OctopusDeploy/Issues#4270]
# Problem
Currently when creating a step template that uses a package, we require the user to define the feed and package directly in the step. The consuming project will then just use whatever package was defined down in the template. This approach has several limitations when users want to provide custom packages for different projects.
* The user is required to use variables to define the package. This is unwieldy, and brittle (and just not aesthetically pleasing 😄 )
* Using variables means that [retention policies](https://octopus.com/docs/administration/retention-policies) can no longer clean up their packages, causing space issues that need to be manually managed by the user outside of Octopus.
* [Auto release creation](https://octopus.com/docs/deployment-process/steps/automatic-release-creation) can no longer trigger since it has no deterministic way to work out if a given package is relevant or not.

A highly voted UV suggestion [Automatic release creation to allow packages with variables in IDs](http://octopusdeploy.uservoice.com/forums/170787/suggestions/6931697) (192 votes, in the top 15) describes the side-effects of this problem.

It feels as though much of the root cause of other trending issues such as 
* [Only retain dynamically referenced packages where the version is used in a release](https://github.com/OctopusDeploy/Issues/issues/4178)
* [Improve logging around retention policies for releases using variables for packages](https://github.com/OctopusDeploy/Issues/issues/3525)

may also be partly mitigated if users had less need to use variables for packages.

# A New Hope
A potential solution for this may to be to provide a flag to optionally source the package in a step template from the consuming project (or consuming step template in the case of Composite Templates 😉 )
![image](https://user-images.githubusercontent.com/1830666/35712716-38d211ce-080f-11e8-9d36-0851a0dc1046.png)

When this step is included in a project, then alongside the parameter that the user can provide, is the package selector. From that point on from Octopus point of view the selected package can be treated like any other package step.

Ultimately in Octopus, packages are at the core of most deployments and so we should ensure using them are treated in as first-class a way as possible.

**Comment by ccamburn** [comment]
I cannot +1 this enough. This is a large reason we cannot use many step templates. A prime example is IIS; we always deploy our sites in a similar manner, but they use different packages. If we use a template, our deployment process has #{variable_name} all over, and the retention policy breaks.

**Comment by blakeduffey** [comment]
Please add me to this list.  It would be very nice to have a cookie cutter Deploy to IIS template where I could specify the package later.

**Comment by TCGDrew** [comment]
+1 x a billion!

**Comment by mildsar** [comment]
When will it be implemented?

**Comment by Flamage82** [comment]
@mildsar This is definitely something we'd like to work on sooner rather than later, and the number of votes in UserVoice clearly shows that our community really cares about this problem.  We think we'll need to address OctopusDeploy/Issues#4390 first before we can tackle this item, so unfortunately we can't give any sort of timeline at this stage.

**Comment by ccamburn** [comment]
Another use case, now that [Octopus supports multiple packages when running scripts](https://octopus.com/blog/script-step-packages):

We often have a "Scripts" package, that contains all scripts used for deployment of all applications, and a "Client" package, that contains custom client sql scripts. We would love to have a single step that executes a script from the Scripts package and references custom-client code in the Client package, but we use Step Templates for most things. This means we cannot use the built-in retention policies if we move to this new model. For now, we have to deploy the client package to the server, then execute the code, then delete the client package.

This would definitely make life easier by having a single step contain all components.

**Comment by jess-ross** [comment]
For your feedback, we've put together a couple of prototypes of how this could work.

One thing to note is the we'd like to introduce a new pattern to the `SelectWithAddRefresh` component. Rather than going to a new browser tab to create a parameter this would happen in place via a dialog. This way the user is kept in the context of the step tab. Alternatively the user can create the Package Parameter on the Parameter tab and select it from the drop down selector on the Step tab. 

_**When viewing the prototypes, if you click on the screen an orange highlight will show you where to click to move through the prototype_

**Deploy package step template :** [https://sketch.cloud/s/egpWq/LvVJLL/play](https://sketch.cloud/s/egpWq/LvVJLL/play)

 **Script step template with reference packages :** [https://sketch.cloud/s/egpWq/g83rlM/play](https://sketch.cloud/s/egpWq/g83rlM/play)

**Comment by ccamburn** [comment]
@jess-ross Looks perfect. I especially like that on the Step Template it specifically provides the option to "Source package from project." That makes it clear and distinct from a normal parameter.

In that scenario, in adding the step template to a project, I would like to request still receiving the appropriate auto-fill and dropdown to select packages from the internal feed, just like when adding the "Deploy a package" step. This way I don't have to type out the entire package name.

**Comment by droyad** [comment]
I don't agree on the point re package retention. In almost all cases the variables are applied successfully and the packages cleaned up.

**Comment by zentron** [comment]
This still feels like a high valued change to me. Will re-investigate in the future

**Comment by ccamburn** [comment]
@zentron Great to hear. Been excited ever since the mockup.

**Comment by hnrkndrssn** [comment]
Release notes: Allow deferring the package selection until the step template is added to a project
